fs/vfs/fs_fchstat.c

```diff
@@ -0,0 +1,286 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <sys/stat.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <nuttx/fs/fs.h>
+
+#include "inode/inode.h"
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: fchstat
+ ****************************************************************************/
+
+static int fchstat(int fd, FAR struct stat *buf, int flags)
+{
+  FAR struct file *filep;
+  int ret;
+
+  /* First, get the file structure.  Note that on failure,
+   * fs_getfilep() will return the errno.
+   */
+
+  ret = fs_getfilep(fd, &filep);
+  if (ret < 0)
+    {
+      goto errout;
+    }
+
+  /* Perform the fchstat operation */
+
+  ret = file_fchstat(filep, buf, flags);
+
+  /* Check if the fchstat operation was successful */
+
+  if (ret >= 0)
+    {
+      /* Successfully fchstat'ed the file */
+
+      return OK;
+    }
+
+errout:
+  set_errno(-ret);
+  return ERROR;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: file_fchstat
+ *
+ * Description:
+ *   file_fchstat() is an internal OS interface. It is functionally similar
+ *   to the combination of fchmod/fchown/futimens standard interface except:
+ *
+ *    - It does not modify the errno variable,
+ *    - It is not a cancellation point,
+ *    - It does not handle socket descriptors, and
+ *    - It accepts a file structure instance instead of file descriptor.
+ *
+ * Input Parameters:
+ *   filep  - File structure instance
+ *   buf    - The stat to be modified
+ *   flags  - The vaild field in buf
+ *
+ * Returned Value:
+ *   Upon successful completion, 0 shall be returned. Otherwise, the
+ *   negative errno shall be returned to indicate the error.
+ *
+ ****************************************************************************/
+
+int file_fchstat(FAR struct file *filep, FAR struct stat *buf, int flags)
+{
+  FAR struct inode *inode;
+  int ret;
+
+  DEBUGASSERT(filep != NULL);
+
+  /* Get the inode from the file structure */
+
+  inode = filep->f_inode;
+  DEBUGASSERT(inode != NULL);
+
+  /* Adjust and check buf and flags */
+
+  if ((flags & CH_STAT_MODE) && (buf->st_mode & ~07777))
```

Comment:
I think this test could always return true: (buf->st_mode & ~07777)
st_mode is of type mode_t that is 16-bit, but 07777 octal value is only 12-bit long, so if the upper 4 bits from st_mode are 0 the resulting value will be different of zero. Did you check it?

Reply:
the up 4bits is used to repesent file type and can not be modified by caller, this check ensure it.